python/ray/serve/_private/router.py

```diff
@@ -1001,11 +1006,38 @@ def __init__(
             call_in_event_loop=event_loop,
         )
 
-        self.metrics_pusher = MetricsPusher()
+        # For autoscaling deployments.
         self.autoscaling_config = None
+        # Track queries sent to replicas for the autoscaling algorithm.
+        self.num_queries_sent_to_replicas = defaultdict(int)
+        self._queries_lock = threading.Lock()
+        # Regularly aggregate and push autoscaling metrics to controller
+        self.metrics_pusher = MetricsPusher()
+        self.metrics_store = InMemoryMetricsStore()
         self.push_metrics_to_controller = controller_handle.record_handle_metrics.remote
 
-    def update_autoscaling_config(self, autoscaling_config):
+    def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+        self._replica_scheduler.update_running_replicas(running_replicas)
+
+        # Prune list of replica ids in self.num_queries_sent_to_replicas
+        # The purpose of this is two-fold:
+        # 1. Only requests sent to RUNNING replicas should be considered
+        #    for the autoscaling algorithm.
```

Comment:
Shouldn't this pruning happen when we send the metrics instead of only when the set is updated? Or is it happening in both places?

Reply:
Yeah, it can. The reason I didn't add it to the controller side is because whenever the list of running replicas change on the controller, handles should be updated as well. By pruning on the handle side, we don't have to send over a map of (replica_id -> request count) to the controller for every handle (and each map will have to be stored in the controller as well). I did think theoretically handles will be slightly more delayed and we can run into out of sync issues, just didn't know how much that would practically impact autoscaling metrics.

Reply:
It's very hard to reason about how this will behave due to ordering issues.

For example, we could have a query that's mid-scheduling, then `update_running_replicas` happens, then we finish scheduling. In this case, we'll end up with very strange behavior -- the replica will be reported as having `1` request ongoing (because in the codepath where we increment the `defaultdict` will return `0` and we'll increment to `1`) until a new `update_running_replicas` occurs.